scale-out-relay/src/main/scala/com/normation/plugins/scaleoutrelay/ScaleOutRelayService.scala

```diff
@@ -0,0 +1,274 @@
+import com.normation.rudder.domain.policies._
+import com.normation.rudder.domain.queries._
+import com.normation.rudder.repository.WoDirectiveRepository
+import com.normation.rudder.repository.WoNodeGroupRepository
+import com.normation.rudder.repository.WoNodeRepository
+import com.normation.rudder.repository.WoRuleRepository
+import com.normation.rudder.rule.category.RuleCategoryId
+import com.normation.rudder.services.nodes.NodeInfoService
+import com.normation.rudder.services.servers.PolicyServerManagementService
+import com.normation.utils.StringUuidGenerator
+import com.softwaremill.quicklens._
+import zio.ZIO
+
+
+
+class ScaleOutRelayService(
+    nodeInfosService              : NodeInfoService
+  , woLDAPNodeGroupRepository     : WoNodeGroupRepository
+  , woLDAPNodeRepository          : WoNodeRepository
+  , woDirectiveRepository         : WoDirectiveRepository
+  , woRuleRepository              : WoRuleRepository
+  , uuidGen                       : StringUuidGenerator
+  , policyServerManagementService : PolicyServerManagementService
+) {
+  val SYSTEM_GROUPS = "SystemGroups"
+  val DISTRIBUTE_POLICY = "distributePolicy"
+  val COMMON = "common"
+
+  def promoteNodeToRelay(uuid: NodeId, actor: EventActor, reason:Option[String]): ZIO[Any, RudderError, NodeInfo] = {
+    val modId = ModificationId(uuidGen.newUuid)
+    for {
+      nodeInfos   <- getNodeToPromote(uuid)
+      updatedNode =  NodeToPolicyServer(nodeInfos)
+      _           <- woLDAPNodeRepository.deleteNode(nodeInfos.node, modId, actor, reason).chainError(s"Remove ${nodeInfos.node.id.value} to update it to policy server failed")
+      newRelay    <- createRelayFromNode(updatedNode, modId, actor, reason).chainError(s"Promote ${nodeInfos.node.id} to relay failed")
+    } yield {
+      newRelay
+    }
+  }
+
+  private[scaleoutrelay] def createRelayFromNode(nodeInf: NodeInfo, modId: ModificationId, actor: EventActor, reason:Option[String])= {
+    for {
+      commonDirective <- createCommonDirective(nodeInf).toIO
+
+      directDistribPolicy = createDirectiveDistributePolicy(nodeInf.node.id)
+      ruleTarget = createPolicyServer(nodeInf.node.id)
+      nodeGroup = createNodeGroup(nodeInf.node.id)
+      ruleDistribPolicy = createRuleDistributePolicy(nodeInf.node.id)
+      ruleSetup = createRuleSetup(nodeInf.node.id)
+
+      categoryId = NodeGroupCategoryId(SYSTEM_GROUPS)
+      activeTechniqueId = ActiveTechniqueId(DISTRIBUTE_POLICY)
+      activeTechniqueIdCommon = ActiveTechniqueId(COMMON)
+
+      _ <- woLDAPNodeRepository.createNode(nodeInf.node, modId, actor, reason).catchAll { err =>
+             val e = Chained(s"Trying to restore initial node configuration", err)
+             PluginLogger.error(e)
+             val old = PolicyServerToNode(nodeInf)
+             woLDAPNodeRepository.createNode(old.node, modId, actor, reason).chainError(s"Rollback promote to relay failed : restaure ${nodeInf.node} configuration failed")
```

Comment:
I think the node Deletion should be the first thing you do here, and not do it in the other function

in fact it should be in one function i think